### PR TITLE
Refactoring app/services/articles/feed.rb

### PR DESCRIPTION
**Pull request by luchiago**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Documentation Update

## Related Tickets & Documents
#3739 

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

### CODE REVIEW

**Comment by citizen428** [comment]
Thanks @luchiago! 

Personally, I'm not a fan of this approach in general. While it may make CodeClimate happy, moving things to modules and re-including them doesn't reduce the complexity of the original code, it just distributes it over several files. On top of that, the extracted methods rely on instance variables that need to be present, which means they are strongly coupled to a single call site / use case anyway.

I see two options:

1. Ignore CodeClimate for the time being for this specific file.
2. Properly extract the functionality into services and explicitly pass in all needed data without relying on specific instance variables being present.

**Comment by luchiago** [comment]
@citizen428 @rhymes I appreciate you guys review. I will move to services and follow your alert.

**Comment by citizen428** [comment]
Thanks for help @luchiago 👍 

Please re-request reviews from us when you're doing with the refactoring.

**Comment by citizen428** [comment]
@luchiago I edited the title of your PR and prefixed it with [WIP] so people realize it's still worked on. Please change it back when it's ready to be reviewed.

File: app/services/articles/array_helper.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+module Articles
+  module ArrayHelper
```
Comment (citizen428) [inline app/services/articles/array_helper.rb:2]: I'm not a big fan of this module, it's only used in one place. Instead of moving this here, how about removing both methods in `feed.rb` and replacing them with

```ruby
def shuffle_results(results)
  first_half = results.first(results.length / 2)
  last_half = results.drop(first_half.size)
  first_half.shuffle + last_half.shuffle
```

Then you can replace all occurrences of `ArrayHelper.first_half(results).shuffle + ArrayHelper.last_half(results)` with `shuffle_results(results)` or even get rid of the intermediate variables, i.e.

```ruby
shuffle_results(more_experience_level_weight_experiment)
```

Comment (rhymes) [reply app/services/articles/array_helper.rb:2]: I agree with @citizen428, I don't tyink we need a helper (which in the context of Rails it should be a view helper) just for splitting an array in two

File: app/services/articles/ranked_and_sorted.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
+module Articles
+  class RankedAndSorted
```
Comment (rhymes) [inline app/services/articles/ranked_and_sorted.rb:2]: I wonder if these new files shouldn't be in `Articles::Feed::RankedAndSorted`

cc. @joshpuetz

Comment (joshpuetz) [reply app/services/articles/ranked_and_sorted.rb:2]: Yes, agreed @rhymes: that would fit our general naming patterns a bit better, and convey that this module is related to the _feed_ specifically

File: app/services/articles/score.rb [hunk 3]
```diff
@@ -0,0 +1,76 @@
+module Articles
+  class Score
```
Comment (joshpuetz) [inline app/services/articles/score.rb:2]: Also would like to see this move to something like `Articles::Feed::Score`

File: spec/services/articles/feed_spec.rb [hunk 4]
```diff
@@ -180,302 +180,6 @@
...
-    it "returns articles in scored order" do
-      allow(feed).to receive(:score_single_article).with(article1).and_return(1)
-      allow(feed).to receive(:score_single_article).with(article2).and_return(2)
-      allow(feed).to receive(:score_single_article).with(article3).and_return(3)
-
-      expect(feed.rank_and_sort_articles(articles)).to eq [article3, article2, article1]
-    end
-  end
-
-  describe ".globally_hot_articles" do
```
Comment (joshpuetz) [inline spec/services/articles/feed_spec.rb:433]: These tests are missing from the refactor: they cover important edge cases and should be included (they may need to become integration level tests now that these classes have been split up

Comment (luchiago) [reply spec/services/articles/feed_spec.rb:433]: How to turn them integration level tests? I removed this test `.globally_hot_articles` because I turned it a private method because it's only used by one method inside the class.

**Comment by luchiago** [comment]
@citizen428 @rhymes @joshpuetz  Alright guys, thanks for reviewing. I'll do the requested changes.

File: spec/services/articles/feed_spec.rb (hunk 4 above)
Comment (luchiago) [reply spec/services/articles/feed_spec.rb:433]: @joshpuetz any thoughts?

**Comment by mstruve** [comment]
Hi @luchiago!

Thanks so much for the PR! Since there has been no activity on this PR for a couple of months and there are some merge conflicts I am going to close it for now. Please feel free to update it, make the requested changes, and reopen it again if you would like!
